src/intersection_helper.jl

```diff
@@ -0,0 +1,49 @@
+"""
+    get_constrained_lowdimset(cpa::CartesianProductArray{N, S}, P::HPolyhedron{N}) where {N<:Real, S<:LazySet{N}}
+
+Preprocess step for intersection between cartesian product array and H-polyhedron.
+Returns low-dimensional HPolytope in constrained dimensions of original cpa,
+constrained variables and variables in corresponding blocks, original block structure
+of low-dimensional set and list of constrained blocks.
+
+### Input
+
+- `cpa` -- Cartesian product array of convex sets
+- `P` -- H-polyhedron
+
+### Output
+
+A tuple of low-dimensional set, list of constrained dimensions, original block
+structure of low-dimensional set and corresponding blocks indices.
+"""
+function get_constrained_lowdimset(cpa::CartesianProductArray{N, S}, P::HPolyhedron{N}) where {N<:Real, S<:LazySet{N}}
+
+    if isbounded(P)
+        blocks, non_empty_length = block_to_dimension_indices(cpa)
+    else
+        constrained_vars = constrained_dimensions(P)
+        blocks, non_empty_length = block_to_dimension_indices(cpa, constrained_vars)
+    end
+
+    array = Vector{S}()
+    sizehint!(array, non_empty_length)
+    vars = Vector{Int}()
+    block_structure = Vector{UnitRange{Int}}()
+    sizehint!(block_structure, non_empty_length)
+
+    last_var = 1
+    for i in 1:length(blocks)
+        start_index, end_index = blocks[i]
+        block_end = last_var + end_index - start_index
+        if start_index != -1
+            push!(array, cpa.array[i])
+            append!(vars, start_index : end_index)
+            push!(block_structure, last_var : block_end)
+        end
+        last_var = block_end + 1
+    end
+
+    cpa_low_dim = HPolytope(constraints_list(CartesianProductArray(array)));
```

Comment:
The following is much cheaper:
```suggestion
    cpa_low_dim = CartesianProductArray(array)
```
However, it will create infinite loops because the callers would call themselves again. Still, I find it cleaner if this helper method really just returns a CPA and the callers do the conversion to `HPolytope` so it is more explicit.

This also requires that the two main methods (`overapproximate` and `is_intersection_empty`) receive a polytopic CPA (at least in the constrained blocks). Please add `polytopic` to the `Input` documentations.

Reply:
> `However, it will create infinite loops because the callers would call themselves again.`

I think `overapproximate` shouldn't call itself again, since we can call it without `CartesianProductArray` type as an argument, but we will get this problem for `is_intersection_empty.jl`

Reply:
> I think overapproximate shouldn't call itself again, since we can call it without CartesianProductArray type as an argument

I was referring to the version *after* the change. There the first argument would be a CPA again. But indeed you call `intersection` instead of `overapproximate`, so this would not be a loop. But there is another problem: We do not have `intersection(::CPA, ::AbstractPolyhedron)` :smile: